db/rocksdb/db.go

```diff
@@ -0,0 +1,428 @@
+package rocksdb
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync"
+	"sync/atomic"
+
+	dbm "github.com/cosmos/cosmos-sdk/db"
+	dbutil "github.com/cosmos/cosmos-sdk/db/internal"
+	"github.com/tecbot/gorocksdb"
+)
+
+var (
+	checkpointFileFormat string = "%020d.db"
+)
+
+var (
+	_ dbm.DBConnection = (*RocksDB)(nil)
+	_ dbm.DBReader     = (*dbTxn)(nil)
+	_ dbm.DBWriter     = (*dbWriter)(nil)
+	_ dbm.DBReadWriter = (*dbWriter)(nil)
+)
+
+// RocksDB is a connection to a RocksDB key-value database.
+type RocksDB = dbManager
+
+type dbManager struct {
+	current *dbConnection
+	dir     string
+	opts    dbOptions
+	vmgr    *dbm.VersionManager
+	mtx     sync.RWMutex
+	// Track open DBWriters
+	openWriters int32
+	cpCache     checkpointCache
+}
+
+type dbConnection = gorocksdb.OptimisticTransactionDB
+
+type checkpointCache struct {
+	cache map[uint64]*cpCacheEntry
+	mtx   sync.RWMutex
+}
+
+type cpCacheEntry struct {
+	cxn      *dbConnection
+	txnCount uint
+}
+
+type dbTxn struct {
+	txn     *gorocksdb.Transaction
+	mgr     *dbManager
+	version uint64
+}
+type dbWriter struct{ dbTxn }
+
+type dbOptions struct {
+	dbo *gorocksdb.Options
+	txo *gorocksdb.OptimisticTransactionOptions
+	ro  *gorocksdb.ReadOptions
+	wo  *gorocksdb.WriteOptions
+}
+
+// NewDB creates a new RocksDB key-value database with inside the given directory.
+// If dir does not exist, it will be created.
+func NewDB(dir string) (*dbManager, error) {
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return nil, err
+	}
+
+	// default rocksdb option, good enough for most cases, including heavy workloads.
+	// 1GB table cache, 512MB write buffer(may use 50% more on heavy workloads).
+	// compression: snappy as default, need to -lsnappy to enable.
+	bbto := gorocksdb.NewDefaultBlockBasedTableOptions()
+	bbto.SetBlockCache(gorocksdb.NewLRUCache(1 << 30))
+	bbto.SetFilterPolicy(gorocksdb.NewBloomFilter(10))
+	dbo := gorocksdb.NewDefaultOptions()
+	dbo.SetBlockBasedTableFactory(bbto)
+	dbo.SetCreateIfMissing(true)
+	dbo.IncreaseParallelism(runtime.NumCPU())
+	// 1.5GB maximum memory use for writebuffer.
+	dbo.OptimizeLevelStyleCompaction(512 * 1024 * 1024)
```

Comment:
This value doesn't match the comment above. 1.5GiB isn't 512 * 1024 * 1024. I'd simplify it and make it succinct as:
1<<30 + 512<<20

Reply:
Good catch